cmd/lotus-shed/gas-estimation.go

```diff
@@ -0,0 +1,285 @@
+		if err != nil {
+			return err
+		}
+
+		tw := tabwriter.NewWriter(os.Stdout, 8, 2, 2, ' ', tabwriter.AlignRight)
+		res, err := sm.CallAtStateAndVersion(ctx, msg, []types.ChainMsg{}, executionTs, stateRootCid, network.Version(nv))
+		if err != nil {
+			return err
+		}
+		fmt.Println("Total gas used: ", res.MsgRct.GasUsed)
+		printInternalExecutions(0, []types.ExecutionTrace{res.ExecutionTrace}, tw)
+
+		return tw.Flush()
+	},
+}
+
+var replayOfflineCmd = &cli.Command{
+	Name:        "replay-offline",
+	Description: "replay a message on the specified stateRoot and network version to get an execution trace",
+	ArgsUsage:   "[messageCid]",
+	Flags: []cli.Flag{
+		&cli.StringFlag{
+			Name:  "repo",
+			Value: "~/.lotus",
+		},
+		&cli.Int64Flag{
+			Name:  "lookback-limit",
+			Value: 10000,
+		},
+	},
+	Action: func(cctx *cli.Context) error {
+		ctx := context.TODO()
+
+		if cctx.NArg() != 1 {
+			return lcli.IncorrectNumArgs(cctx)
+		}
+
+		messageCid, err := cid.Decode(cctx.Args().Get(0))
+		if err != nil {
+			return fmt.Errorf("failed to parse input: %w", err)
+		}
+
+		lookbackLimit := cctx.Int("lookback-limit")
+
+		fsrepo, err := repo.NewFS(cctx.String("repo"))
+		if err != nil {
+			return err
+		}
+
+		lkrepo, err := fsrepo.Lock(repo.FullNode)
+		if err != nil {
+			return err
+		}
+
+		defer lkrepo.Close() //nolint:errcheck
+
+		bs, err := lkrepo.Blockstore(ctx, repo.UniversalBlockstore)
+		if err != nil {
+			return fmt.Errorf("failed to open blockstore: %w", err)
+		}
+
+		defer func() {
+			if c, ok := bs.(io.Closer); ok {
+				if err := c.Close(); err != nil {
+					log.Warnf("failed to close blockstore: %s", err)
+				}
+			}
+		}()
+
+		mds, err := lkrepo.Datastore(context.Background(), "/metadata")
+		if err != nil {
+			return err
+		}
+
+		dcs := build.DrandConfigSchedule()
+		shd := beacon.Schedule{}
+		for _, dc := range dcs {
+			bc, err := drand.NewDrandBeacon(1598306400, build.BlockDelaySecs, nil, dc.Config) // 1598306400 is mainnet genesis time
+			if err != nil {
+				return xerrors.Errorf("creating drand beacon: %w", err)
+			}
+			shd = append(shd, beacon.BeaconPoint{Start: dc.Start, Beacon: bc})
+		}
+
+		cs := store.NewChainStore(bs, bs, mds, filcns.Weight, nil)
+		defer cs.Close() //nolint:errcheck
+
+		sm, err := stmgr.NewStateManager(cs, filcns.NewTipSetExecutor(), vm.Syscalls(ffiwrapper.ProofVerifier), filcns.DefaultUpgradeSchedule(), shd)
+		if err != nil {
+			return err
+		}
+
+		msg, err := cs.GetMessage(ctx, messageCid)
+		if err != nil {
+			return err
+		}
+
+		err = cs.Load(ctx)
+		if err != nil {
+			return err
+		}
+
+		ts, _, _, err := sm.SearchForMessage(ctx, cs.GetHeaviestTipSet(), messageCid, abi.ChainEpoch(lookbackLimit), true)
+		if err != nil {
+			return err
+		}
+		if ts == nil {
+			return xerrors.Errorf("could not find message within the last %d epochs", lookbackLimit)
+		}
+		executionTs, err := cs.GetTipsetByHeight(ctx, ts.Height()-2, ts, true)
```

Comment:
explain why -2 pls

Reply:
That is probably question for me, no idea, that is what worked at the time :P